File: src/javascript/algorithms/simulation/reconstructItenary/reconstructItenary.js
// Reconstruct Itinerary (LC 332)
// DFS + Hierholzer's algorithm  TC: O(E log E)  SC: O(E)

/**
 * @param {string[][]} tickets
 * @return {string[]}
 */
function reconstructItinerary(tickets) {
    const graph = {};
    for (const [src, dst] of tickets) {
        if (!graph[src]) graph[src] = [];
        graph[src].push(dst);
    }
    for (const src in graph) {
        graph[src].sort().reverse(); // descending so pop() yields smallest
    }

    const result = [];

    function dfs(airport) {
        while (graph[airport] && graph[airport].length > 0) {
            dfs(graph[airport].pop());
        }
        result.push(airport);
    }

    dfs("JFK");
    return result.reverse();
}

function runTests() {
    const tests = [
        {
            tickets: [["MUC","LHR"],["JFK","MUC"],["SFO","SJC"],["LHR","SFO"]],
            expected: ["JFK","MUC","LHR","SFO","SJC"],
        },
        {
            tickets: [["JFK","SFO"],["JFK","ATL"],["SFO","ATL"],["ATL","JFK"],["ATL","SFO"]],
            expected: ["JFK","ATL","JFK","SFO","ATL","SFO"],
        },
        {
            tickets: [["JFK","KUL"],["JFK","NRT"],["NRT","JFK"]],
            expected: ["JFK","NRT","JFK","KUL"],
        },
        {
            tickets: [["JFK","A"],["A","JFK"]],
            expected: ["JFK","A","JFK"],
        },
    ];

    for (const { tickets, expected } of tests) {
        const result = reconstructItinerary(tickets);
        const pass = JSON.stringify(result) === JSON.stringify(expected);
        console.log(`[${pass ? "PASS" : "FAIL"}] result=${JSON.stringify(result)} expected=${JSON.stringify(expected)}`);
    }
}

runTests();

File: src/python/algorithms/simulation/reconstructItenary/reconstruct_itenary.py
# Reconstruct Itinerary (LC 332)
# DFS + Hierholzer's algorithm  TC: O(E log E)  SC: O(E)

from typing import List


def reconstruct_itinerary(tickets: List[List[str]]) -> List[str]:
    graph = {}
    for src, dst in tickets:
        graph.setdefault(src, []).append(dst)
    for src in graph:
        graph[src].sort(reverse=True)  # descending so pop() yields smallest

    result = []

    def dfs(airport: str) -> None:
        while graph.get(airport):
            dfs(graph[airport].pop())
        result.append(airport)

    dfs("JFK")
    return result[::-1]


if __name__ == "__main__":
    test_cases = [
        (
            [["MUC","LHR"],["JFK","MUC"],["SFO","SJC"],["LHR","SFO"]],
            ["JFK","MUC","LHR","SFO","SJC"],
        ),
        (
            [["JFK","SFO"],["JFK","ATL"],["SFO","ATL"],["ATL","JFK"],["ATL","SFO"]],
            ["JFK","ATL","JFK","SFO","ATL","SFO"],
        ),
        (
            [["JFK","KUL"],["JFK","NRT"],["NRT","JFK"]],
            ["JFK","NRT","JFK","KUL"],
        ),
        (
            [["JFK","A"],["A","JFK"]],
            ["JFK","A","JFK"],
        ),
    ]

    for tickets, expected in test_cases:
        result = reconstruct_itinerary(tickets)
        status = "PASS" if result == expected else "FAIL"
        print(f"[{status}] result={result} expected={expected}")
